### ISSUE TITLE
structural search directly uses zoekt API

### ISSUE THREAD

**Issue by keegancsmith** [sourcegraph/sourcegraph#14619]
This issue is about discussing (and possibly PoCing) structural searching being more tightly integrated with zoekt. We have a more scaleable structural search as a search roadmap item, this is one possible project to help achieve that.

### Current

1. Convert query `Q` into regex `R` that matches a superset of documents
2. Query zoekt with `R`, but only asking for file path matches.
3. Send the list of file path matches as a list to constrain search in `searcher`
4. `searcher` fetches the archive for each repo@commit for a file path, then runs comby.

### Proposed

First we need `comby` to add some sort of API/RPC/etc over stdin/stdout or a unix socket. This can be an API where we send commands like `query: ..., name: foo.go, content: ...` and comby returns a response. Then

1. Convert query `Q` into regex `R` that matches a superset of documents.
2. Query zoekt with `R`, but only ask for file path matches *AND* for zoekt to return file contents.
3. As zoekt returns results, we send them as commands to comby.

This avoids the need to fetch an archive, or extra complexity around state management. This is purely a bit of glue between zoekt and a pool of comby processes. With a streaming API on top of zoekt, this would work quite pleasantly.

Alternatively if we have this sort of API to comby, one can imagine calling out to comby in Zoekt. One can imagine a very cool command would instead be something like `query: ..., name: foo.go, file: /path/to/shard, offset: 1234, length: 100`. This allows comby to just mmap in the content zoekt has as a shard, and avoids needing to buffer stuff around.

cc @sourcegraph/search

**Comment by keegancsmith** [comment]
@rvantonder how reasonable is the proposed API addition to comby? This is somewhat inspired by how we interact with ctags over stdin/stdout. If we wanted to introduce some extra tooling, one can imagine using something like flatbuffers/etc, except the API is so simple I'd rather do nuts and bolts sort of marshalling (eg `write([]byte(len(b))); write(b)`)

**Comment by rvantonder** [comment]
I think the proposal is worth trying. I have in mind to try a direction that decouples a bit from Zoekt, but it is different/orthogonal enough that I'm not going to import those concerns here. 

> First we need comby to add some sort of API/RPC/etc over stdin/stdout or a unix socket. 
     
  The Go library wraps the command line already with most of what's needed, the only thing that we need to add is an option to pipe file contents to stdin, if I understand this correctly? So I could expand what the [Input](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/comby/types.go#L15) arg to accept file content, through a buffer that's piped to stdin?

I can add that ~easily. For the rest of this proposal, my main question is--are you proposing to hack on this, or something you're proposing I should pick up? I have in mind to try the decoupling above (which starts by asking Zoekt only for repos instead of files, or operating without Zoekt altogether for, e.g,. unindexed paths) before exploring this "tighter integration" proposal.

**Comment by keegancsmith** [comment]
> > First we need comby to add some sort of API/RPC/etc over stdin/stdout or a unix socket.
> 
> The Go library wraps the command line already with most of what's needed, the only thing that we need to add is an option to pipe file contents to stdin, if I understand this correctly? So I could expand what the [Input](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/comby/types.go#L15) arg to accept file content, through a buffer that's piped to stdin?

The idea is to avoid the cost of constantly starting up comby. Could you support a tarball over stdin that you stream and search? That seems like a reasonable API that is not sourcegraph specific.

> are you proposing to hack on this, or something you're proposing I should pick up?

I'm not sure who or when we will hack on this. I was just front loading this proposal for when we start tackling structural search performance. The reason I like this approach is it avoids the need to fetch archives first.

**Comment by rvantonder** [comment]
For the monorepo case, I think the archive copying is especially bad (it would ask for everything from gitserver right?). Because this basically breaks structural search, I'm going to spend some time looking into a low-effort way of zipping up only the file contents and sending only that over to `searcher` for `comby` to search over, skipping a fetch of the entire archive. This should at least unbreak the monorepo timeout issues ([context](https://sourcegraph.slack.com/archives/CHEKCRWKV/p1604265162383600)). So basically, similar to the query you're describing above (get contents from Zoekt), but without the streaming (instead just package it up in a comparably ~small zip and send over--this should be OK to go in the frontend, maybe?) LMK your thoughts.

Medium term (or whenever soonest), I can pick up adding a streaming/tar endpoint for comby and/or searcher, I just think that's going to take a bit longer and there are more concerns here, and it feels like it may be dependent on your WIP streaming work too.

**Comment by keegancsmith** [comment]
Zoekt's API supports asking for file contents of matches. I suppose streaming is necessary otherwise I can imagine its easy to OOM. Sending the file contents to searcher from frontend sounds like a good general API. :+1:

**Comment by rvantonder** [comment]
Hmmm good point about the OOM possibility. I think streaming directly to searcher/comby is going to be tricky to do quickly (thinking I want to spend about 3 days on a quicker fix). To counter the OOM problem, what I think works is to stream contents from frontend and just write to disk (uncompressed zip archive) on the searcher side and then run comby on that. This would lower the scope of work and still allow comby to still run in parallel. 

Once it's possible to stream directly to comby, we can add orchestration (e.g., managing pool of workers) to pipe that stream input directly, and then look at moving that work to Zoekt directly (which is the original proposal this issue started off with :-))

**Comment by rvantonder** [comment]
> Sending the file contents to searcher from frontend sounds like a good general API. 👍

I've been grappling with this some more and just had a chat with @stefanhengl to confirm another direction that might simplify things. Basically, it's annoying/debt-prone to orchestrate sending the contents via frontend to searcher, and potentially avoidable: what's your intuition about giving searcher a handle to Zoekt client and querying it directly? What would happen here is:

frontend gets structural search request -> always sends it to searcher -> searcher queries Zoekt for contents

This would avoid the content proxy through frontend (and a lot clearer to me how to get the data), but the unknown here is whether it is at all reasonable/feasible to give a Zoekt client handle to searcher. Thoughts @keegancsmith?

This would, architecturally, also help with two other pieces of complexity:

- removing the structural search logic that queries files in `textsearch.go`, it's very complex
- open up an easy way to do unindexed structural search, because all requests go to the same searcher endpoint (previously, if I wanted to add unindexed structural search, I would have to create a new codepath that doesn't go through the indexed search codepath).

**Comment by keegancsmith** [comment]
@rvantonder that is pretty much what I proposed above (if you ignore streaming), so that sounds good :) Currently Zoekt is configured only in frontend, but the code to do horizontal searches is in `./internal/search` so this sounds very doable. I think the "cleanest" way to do this is for the frontend to communicate to searcher where to find zoekt. This can be done via conf's endpoints part. We already do this for gitserver. Two options on what to communicate:
1. The endpoint URI. This requires searcher access to the k8s API for service discovery. This requires a bunch of infra changes, so I don't think this makes sense.
2. Return the endpoints already discovered by frontend. This seems nice :)

searcher becomes a little less "pure", but this honestly sounds like a great addition. Maybe in the design of the code you can somewhat hide that you are speaking to zoekt in the searcher code, and have a generic interface which just takes in a query and does the necessary magic to get back file contents to search.

**Comment by rvantonder** [comment]
To make this happen needs two major changes:

1. The frontend logic that calls Zoekt and then passes file paths to searcher needs to be removed. It'll be replaced with its own path that calls searcher directly with all the info we need to then call Zoekt directly in searcher, and pipe to comby.

2. Searcher needs the logic and types to call Zoekt directly and independently.

In a separate experiment I partially tackled (1) by finding and removing the logic that calls Zoekt. I'm going to continue with the path to make changes to the frontend (`search_results.go`, `textsearch.go`) and replace it with the logic to call out to searcher.

@camdencheek can start looking into (2). The thing to do is to move the bits in our frontend that call out to Zoekt to get results, and make sure we can do essentially the same thing, but in searcher. I suggest going about it this way for a start, so we can make incremental progress and reduce complexity and having too much stuff to deal with at once:

- create a fresh file `zoekt_search.go` in [searcher](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/tree/cmd/searcher/search). This'll be the code that calls out to Zoekt
- look at this call to [`zoektSearchHEADOnlyFiles`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/zoekt.go#L230). We need to essentially be able to call that inside the `zoekt_search.go` in searcher. To do that, start by copying the contents of that function to `zoekt_search.go` and import/resolve all the dependencies and types, and see if you can get it to compile (this is the part where I'm not sure what we'll run into yet). For now assume that we have the arguments needed to call that function. Part of (1) I'll be working on is how we can get those arguments to searcher (I think the the tricky bit will be figuring out how to either pass this [`Zoekt`](ob/internal/search/types.go#L118) structure over to searcher, or create it).
- We'll need to make some changes to `zoektSearchHEADOnlyFiles`, which currently returns files matching a query--instead we'll want contents to pipe to comby, and a couple of other changes. We can sync again at this point. So feel free to use a different name for that function in `zoekt_search.go` (just `zoektSearch`).

**Comment by rvantonder** [comment]
New code path has structural search talking to Zoekt. Closing this in favor of smaller remaining TODOs to activate the new path.
